### PR TITLE
Fix extra unreachable generation

### PR DESCRIPTION
**Pull request by aheejin**
Currently various expressions handle this differently, and now we
consistently follow this rules:

---
For all non-control-flow value-returning instructions, if a type of an
expression is unreachable, we emit an unreachable and don't emit the
instruction itself. If we don't emit an unreachable, instructions that
follow can have validation failure in wasm binary format. For example:
```
[unreachable] (f32.add
[unreachable]   (i32.eqz
[unreachable]     (unreachable)
                )
                ...
              )
```
This is a valid prgram in binaryen IR, because the unreachable type
propagates out of an expression, making both i32.eqz and f32.add
unreachable. But in binary format, this becomes:
```
unreachable
i32.eqz
f32.add       ;; validation failure; it expects f32 but takes an i32!
```
And here f32.add causes validation failure in wasm validation. So in this
case we add an unreachable to prevent following instructions to consume
the current value (here i32.eqz).

In actual tests, I used `global.get` to an f32 global, which does not
return a value, instead of `f32.add`, because `f32.add` itself will not
be emitted if one of argument is unreachable.

---

So the changes are:
- For instructions that don't return a value, removes unreachable
  emitting code if it exists.
- Add the unreachable emitting code for value-returning instructions if
  there isn't one.
- Check for unreachability only once after emitting all children for
  atomic instructions. Currently only atomic instructions check
  unreachability after visiting each children and bail out right after,
  which is valid, but not consistent with others.
- Don't emit an extra unreachable after a return (and return_call). I
  guess it is unnecessary?

### CODE REVIEW

**Comment by aheejin** [comment]
This currently fails CI because a text printing bug, which will be fixed by #2265.

File: src/wasm-stack.h [hunk 1]
```diff
@@ -382,11 +382,33 @@ void BinaryenIRWriter<SubType>::visitCall(Call* curr) {
   for (auto* operand : curr->operands) {
     visit(operand);
   }
-  emit(curr);
-  // TODO FIXME: this and similar can be removed
-  if (curr->type == unreachable) {
+
+  // For non-control-flow value-returning instructions, if the type of an
+  // expression is unreachable, we emit an unreachable and don't emit the
+  // instruction itself, If we don't emit an unreachable, instructions that
```
Comment (kripken) [inline src/wasm-stack.h]: `,` => `.`

File: src/wasm-stack.h [hunk 2]
```diff
@@ -382,11 +382,33 @@ void BinaryenIRWriter<SubType>::visitCall(Call* curr) {
...
+  // follow can have validation failure in wasm binary format. For example:
```
Comment (kripken) [inline src/wasm-stack.h]: `have` => `have a`

File: src/wasm-stack.h (hunk 1 above)
Comment (aheejin) [reply src/wasm-stack.h]: Done

File: src/wasm-stack.h (hunk 2 above)
Comment (aheejin) [reply src/wasm-stack.h]: Done

**Comment by aheejin** [comment]
`test/unit/input/tail_call_target_feature.wasm` change is to remove an `unreachable` from the input.
The original file, disassembled, is
```wast
(func $foo
  (return_call $bar)
  (unreachable)
)
``` 

When reading binary, once it sees an unreachable type expression (here `return_call`), it removes the `unreachable` at the end. But we used to add an `unreachable` instruction in the binary writer after a `return_call` (because its type is unreachable), it was able to roundtrip by accident. Now we don't emit an `unreachable` after `return_call` anymore, so the output file does not have `unreachable` at the end. So I deleted the `unreachable` from the test input.

**Pull request merged**